### ISSUE TITLE
Impossible hourly stats

### ISSUE THREAD

**Issue by sadaszewski** [NodeBB/NodeBB#10436]
This morning to my surprise I have found the following hourly statistics on the forum:

![image](https://user-images.githubusercontent.com/1378525/160269071-901c7cfe-949f-4caa-bfc1-7662a2bca8b0.png)

The daily stats look correct. How could this be possible? I have recently upgraded from 1.19.1 to 1.19.5 but initially everything was fine and actually following that flattening of the curves as you can see the most recent hourly entries also look fine again. Any idea on what might be the cause or how one would go about diagnosing this behavior? Thank you in advance.

PS. I cannot say what was happening during the night but I can assure that 24 hours from now those entries at 9 AM etc. were definitely different values.

**Comment by sadaszewski** [comment]
Surprisingly, the data in the database for the relevant period does not correspond to what the chart is displaying. It looks very sensible / as expected. Here it is:

`analytics:uniquevisitors|14.0|1648278000000
analytics:uniquevisitors|19.0|1648281600000
analytics:uniquevisitors|12.0|1648285200000
analytics:uniquevisitors|15.0|1648288800000
analytics:uniquevisitors|23.0|1648292400000
analytics:uniquevisitors|13.0|1648296000000
analytics:uniquevisitors|19.0|1648299600000
analytics:uniquevisitors|13.0|1648303200000
analytics:uniquevisitors|15.0|1648306800000
analytics:uniquevisitors|17.0|1648310400000
analytics:uniquevisitors|15.0|1648314000000
analytics:uniquevisitors|12.0|1648317600000
analytics:uniquevisitors|23.0|1648321200000
analytics:uniquevisitors|20.0|1648324800000
analytics:uniquevisitors|23.0|1648328400000
analytics:uniquevisitors|16.0|1648332000000
analytics:uniquevisitors|19.0|1648335600000
analytics:uniquevisitors|15.0|1648339200000
analytics:uniquevisitors|13.0|1648342800000
analytics:uniquevisitors|13.0|1648346400000
analytics:uniquevisitors|16.0|1648350000000
analytics:uniquevisitors|20.0|1648353600000
analytics:uniquevisitors|24.0|1648357200000
analytics:uniquevisitors|14.0|1648360800000`

**Comment by sadaszewski** [comment]
And this is what the analytics REST endpoint returns o_O

```
{"query":{"units":"hours","until":"2022-03-27T06:45:59.392Z","count":"24"},"result":{"uniquevisitors":[13,13,13,13,13,13,13,13,13,13,13,13,13,13,13,13,13,13,13,13,16,20,24,17],"pageviews":[81,81,81,81,81,81,81,81,81,81,81,81,81,81,81,81,81,81,81,159,133,111,126,76],"pageviews:registered":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,32,1],"pageviews:bot":[33,33,33,33,33,33,33,33,33,33,33,33,33,33,33,33,33,33,33,72,71,54,35,24],"pageviews:guest":[48,48,48,48,48,48,48,48,48,48,48,48,48,48,48,48,48,48,48,87,62,57,59,51]}}
```

**Comment by sadaszewski** [comment]
I have verified that the CLI (./nodebb info) gives the same erroneous output.

**Comment by sadaszewski** [comment]
OK I think it is this loop in `analytics.js:215`:

```
for (let i = 0, ii = numHours; i < ii; i += 1) {
	hoursArr.push(hour.getTime());
	hour.setHours(hour.getHours() - 1, 0, 0, 0);
}
```

This line only manages to go back to midnight and then the date is stuck, hence the repeated values. I do not know how it is possible that it seemed to be working fine so far but both in Node and in the browser I can confirm that the date cannot flip by using negative setHours(). Could someone please comment? Thanks.

**Comment by sadaszewski** [comment]
Oh actually I realized that this night there was a DST time change in my region. But it does not change the fact that the dates cannot flip beyond midnight for me...

**Comment by sadaszewski** [comment]
Anyhow, this would seem to fix it for me:

```
for (let i = 0, ii = numHours; i < ii; i += 1) {
	hoursArr.push(hour.getTime() - i * 3600 * 1000);
}
```

**Comment by sadaszewski** [comment]
OK, I looked at it more closely and it's NOT about midnight. It's at the time when the Daylight Saving Time is activated. The `getHours() - 1` logic fails there, as  we get `-1` hours from the code and `+1` hours from the automatic handling of the DST which leaves us with the same value. It's an interesting corner case. Hope I managed to get this clear. The fix above still seems sort of reasonable.
